### PR TITLE
[ODS-4521] Docker Compose file for sandbox deployment

### PR DESCRIPTION
**Pull request by ptar-innive**

### CODE REVIEW

**Review by blmeyers — CHANGES_REQUESTED** [review]

File: DB-Admin/init-database.sh [hunk 1]
```diff
@@ -0,0 +1,17 @@
+#!/bin/bash
+# SPDX-License-Identifier: Apache-2.0
+# Licensed to the Ed-Fi Alliance under one or more agreements.
+# The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
+# See the LICENSE and NOTICES files in the project root for more information.
+set -e
+
+psql -v ON_ERROR_STOP=1 --username "$POSTGRES_USER" --dbname "$POSTGRES_DB" <<-EOSQL
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply DB-Admin/init-database.sh:8]: done

File: DB-ODS/init-database.sh [hunk 2]
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash
+# SPDX-License-Identifier: Apache-2.0
+# Licensed to the Ed-Fi Alliance under one or more agreements.
+# The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
+# See the LICENSE and NOTICES files in the project root for more information.
+
+set -e
+
+psql -v ON_ERROR_STOP=1 --username "$POSTGRES_USER" --dbname "$POSTGRES_DB" <<-EOSQL
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply DB-ODS/init-database.sh:9]: done

File: DB-ODS/init-database.sh [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+    GRANT ALL PRIVILEGES ON DATABASE "EdFi_Ods_Minimal_Template" TO postgres;
+EOSQL
+
+psql -v --no-password --tuples-only --username postgres --dbname "EdFi_Ods_Minimal_Template" --file /app/EdFi_Ods_Minimal_Template.sql
+
+psql -v --no-password --tuples-only --username postgres --dbname "EdFi_Ods_Populated_Template" --file /app/EdFi_Ods_Populated_Template.sql
+
+psql -v ON_ERROR_STOP=1 --username "$POSTGRES_USER" --dbname "$POSTGRES_DB" <<-EOSQL
+    UPDATE pg_database SET datistemplate='true', datallowconn='false' WHERE datname in ('EdFi_Ods_Populated_Template', 'EdFi_Ods_Minimal_Template');
+    CREATE DATABASE "EdFi_ODS" TEMPLATE 'EdFi_Ods_Minimal_Template';
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply DB-ODS/init-database.sh:22]: done

File: DB-ODS/init-database.sh [hunk 4]
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash
+# SPDX-License-Identifier: Apache-2.0
+# Licensed to the Ed-Fi Alliance under one or more agreements.
+# The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
+# See the LICENSE and NOTICES files in the project root for more information.
+
+set -e
+
+psql -v ON_ERROR_STOP=1 --username "$POSTGRES_USER" --dbname "$POSTGRES_DB" <<-EOSQL
+    CREATE DATABASE "EdFi_Ods_Populated_Template" TEMPLATE template0;
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply DB-ODS/init-database.sh:10]: initially the intent is to run sandbox as a "production" environment so I modified the script to support both a shared instance and sandbox env.

File: DB-Sandbox/EdFi_Admin.sql [hunk 5]
```diff
@@ -0,0 +1,1543 @@
+--
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply DB-Sandbox/EdFi_Admin.sql:1]: currently we do not ahve a build that creates sandbox and security as pg dumps. This was generated using the cloud build.

[the remark this answers is not in the document]
Comment (jamessmckay) [reply DB-Sandbox/EdFi_Admin.sql:1]: ODS-4677

File: DB-Sandbox/EdFi_Security.sql [hunk 6]
```diff
@@ -0,0 +1,1287 @@
+--
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply DB-Sandbox/EdFi_Security.sql:1]: ODS-4677

File: DB-Sandbox/init-database.sh [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+# See the LICENSE and NOTICES files in the project root for more information.
+
+set -e
+
+psql ON_ERROR_STOP=1 --username "$POSTGRES_USER" --dbname "$POSTGRES_DB" <<-EOSQL
+    CREATE DATABASE "EdFi_Ods_Minimal_Template" TEMPLATE template0;
+    CREATE DATABASE "EdFi_Ods_Populated_Template" TEMPLATE template0;
+    CREATE DATABASE "EdFi_Admin" TEMPLATE template0;
+    CREATE DATABASE "EdFi_Security" TEMPLATE template0;
+    GRANT ALL PRIVILEGES ON DATABASE "EdFi_Ods_Populated_Template" TO postgres;
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply DB-Sandbox/init-database.sh:14]: done

File: Web-GateWay/Dockerfile [hunk 8]
```diff
@@ -0,0 +1,12 @@
+# SPDX-License-Identifier: Apache-2.0
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply Web-GateWay/Dockerfile:1]: ODS-4679

File: DB-Admin/EdFi_Admin.sql [hunk 9]
```diff
@@ -31,42 +31,42 @@ ALTER SCHEMA dbo OWNER TO postgres;
 
 CREATE FUNCTION dbo.getclientfortoken(accesstoken uuid) RETURNS TABLE(key character varying, usesandbox boolean, studentidentificationsystemdescriptor character varying, educationorganizationid integer, claimsetname character varying, namespaceprefix character varying, profilename character varying, creatorownershiptokenid smallint, ownershiptokenid smallint)
     LANGUAGE plpgsql
-    AS $$
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply DB-Admin/EdFi_Admin.sql:34]: probably due to change of me removing cr-lf, to lf. I will be replacing the files with the windows version, and adding a gitattributes file.

File: DB-Admin/Dockerfile [hunk 10]
```diff
@@ -0,0 +1,23 @@
+# SPDX-License-Identifier: Apache-2.0
+# Licensed to the Ed-Fi Alliance under one or more agreements.
+# The Ed-Fi Alliance licenses this file to you under the Apache License, Version 2.0.
+# See the LICENSE and NOTICES files in the project root for more information.
+
+FROM postgres:11-alpine
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply DB-Admin/Dockerfile:6]: fixed, and updated the docker files to use the fixed scripts. created new tickets to have admin and security dbs created as nuget packages

File: Web-GateWay/edfi-prod.conf [hunk 11]
```diff
@@ -0,0 +1,50 @@
+upstream webapi {
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply Web-GateWay/edfi-prod.conf:1]: (y)

File: Web-GateWay/edfi-stage.conf [hunk 12]
```diff
@@ -19,7 +19,7 @@ server {
 }
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply Web-GateWay/edfi-stage.conf:19]: this should be for a reverse proxy implementation

File: Web-GateWay/nginx.conf [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+http {
+    default_type  application/octet-stream;
+    sendfile        on;
+
+    keepalive_timeout  65;
+
+    server {
+        listen       5001;
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply Web-GateWay/nginx.conf:15]: deferring until we get reverse proxy fully working. ODS-4675

File: Web-Sandbox-Admin/sandbox.appsettings.json [hunk 14]
```diff
@@ -0,0 +1,60 @@
+﻿{
+  "PreserveLoginUrl": "true",
+  "ClientValidationEnabled": "true",
+  "UnobtrusiveJavaScriptEnabled": "true",
+  "DefaultApplicationName": "Default Sandbox Application",
+  "DefaultClaimSetName": "Ed-Fi Sandbox",
+  "SandboxAdminSQLCommandTimeout": "120",
+  "DefaultOperationalContextUri": "uri://ed-fi-api-host.org",
+  "OAuthUrl": "http://api:80/oauth/",
+  "UseReverseProxyHeaders": false,
```
[the remark this answers is not in the document]
Comment (jamessmckay) [reply Web-Sandbox-Admin/sandbox.appsettings.json:10]: agreed ODS-4595

File: DB-ODS/init-database.sh (hunk 4 above)
[the remark this answers is not in the document]
Comment (blmeyers) [reply DB-ODS/init-database.sh:10]: We can address for the production implementation, but I think we're going to need specialized scripts, one for shared, one for sandbox, one for year specific.  We don't want to set up extra databases unless we have to.

**Review by blmeyers — APPROVED** [review]

**Pull request merged**